File: utils/leaderboardCalc.js
const fs = require('fs');
const path = require('path');
const { syncToSite } = require('./syncToSite');
const { DATA_PATHS, WEBSITE_REPO, GITHUB_TOKEN } = require('../utils/constants');

const playersPath = path.join(__dirname, '../', DATA_PATHS.PLAYERS);
const leaderboardPath = path.join(__dirname, '../', DATA_PATHS.LEADERBOARD_WEEKLY);

function calculateLeaderboard() {
  let players;

  try {
    players = JSON.parse(fs.readFileSync(playersPath, 'utf8'));
  } catch {
    players = [];
  }

  const sorted = players
    .sort((a, b) => b.xp - a.xp)
    .map((player, index) => ({
      rank: index + 1,
      name: player.name,
      points: player.xp,
      discordId: player.discordId
    }));

  fs.writeFileSync(leaderboardPath, JSON.stringify(sorted, null, 2));

  try {
    syncToSite('leaderboard_weekly.json', WEBSITE_REPO, GITHUB_TOKEN);
  } catch (err) {
    console.error('❌ Failed to sync leaderboard:', err);
  }
}

module.exports = { calculateLeaderboard };

File: utils/leaderboardCalc.py
import json
from utils.constants import DATA_PATHS, WEBSITE_REPO, GITHUB_TOKEN
from utils.syncToSite import sync_to_site

def calculate_leaderboard():
    players_path = DATA_PATHS["PLAYERS"]
    leaderboard_path = DATA_PATHS["LEADERBOARD_WEEKLY"]

    try:
        with open(players_path, "r", encoding="utf8") as f:
            players = json.load(f)
    except:
        players = []

    sorted_players = sorted(players, key=lambda p: p.get("xp", 0), reverse=True)

    leaderboard = [
        {
            "rank": index + 1,
            "name": player.get("name"),
            "points": player.get("xp"),
            "discordId": player.get("discordId")
        }
        for index, player in enumerate(sorted_players)
    ]

    with open(leaderboard_path, "w", encoding="utf8") as f:
        json.dump(leaderboard, f, indent=2)

    try:
        sync_to_site("leaderboard_weekly.json", WEBSITE_REPO, GITHUB_TOKEN)
    except Exception as e:
        print(f"❌ Failed to sync leaderboard: {e}")
